File: src/content/docs/laromaterial/labb/lab_csharp/lab_01/Lab.cs
/**
 * This module contains lab specific code to assert the functions of the lab 
 * and to calculate the result.
 */
static class Lab
{
    /**
    * Execute the testcase and assert that it was corrent or not and return a
    * status string the can be written out.
    * @param {Function} func - The function to test.
    * @param {Object} expected - The expected result from the function call.
    * @param {Object} args - The arguments to send to the function.
    * @param {number} point - The number of points that this functions is worth in total.
    * @returns {string} A string representation of the test case success or failure.
    */
    static public string Assert(
        Delegate method,
        object[] args,
        object expected,
        int point = 1
    )
    {
        object? result = method.DynamicInvoke(args);
        string success = "❌";

        if (expected == null && result == null)
        {
            success = "✅";
        }
        else if (expected is not null && result is not null && expected.GetType() == result.GetType())
        {
            if (expected.GetType() == typeof(string) && (string)expected == (string)result)
            {
                success = "✅";
            }
            else if (expected.GetType() == typeof(int) && (int)expected == (int)result)
            {
                success = "✅";
            }
            else if (expected.GetType() == typeof(double) && (double)expected == (double)result)
            {
                success = "✅";
            }
            else if (expected.GetType() == typeof(bool) && (bool)expected == (bool)result)
            {
                success = "✅";
            }
        }


        // // Calculate the stats
        // if (!(func.name in stats)) {
        //     stats[func.name] = {
        //     run: 0,
        //     passed: 0,
        //     failed: 0,
        //     point,
        //     }
        // }
        // stats[func.name].run++
        // if (expected === result) {
        //     stats[func.name].passed++
        // } else {
        //     stats[func.name].failed++
        // }

        // // Prepare the argument string
        string argStr = argsAsString(args);
        string expectedStr = expected is not null ? argsAsString([expected]) : "";
        string resultStr = result is not null ? argsAsString([result]) : "";
        string methodName = method.Method.Name;
        string expectedType = expected is not null ? expected.GetType().ToString().Substring(7) : "";
        string resultType = result is not null ? result.GetType().ToString().Substring(7) : "";

        // return `${success} ${stats[func.name].point}p ${func.name}(${argStr}), expected: ${expectedStr} (${typeof expected}), actual: ${resultStr} (${typeof result})`
        return $"{success} {methodName}({argStr}), expected: {expectedStr} ({expectedType}), actual: {resultStr} ({resultType})";
    }

    /**
    * Format a value to its string representation.
    * @param {Object} args - A value to parse and format as a string.
    * @returns {string} A string representation of the value.
    */
    static string argsAsString(object[] args)
    {
        string argStr = "";
        for (int i = 0; i < args.Length; i++)
        {
            if (args[i].GetType() == typeof(string))
            {
                argStr += $"\"{args[i]}\", ";
            }
            // else if (typeof arg === 'object' && Array.isArray(arg)) {
            //     argStr += `[${arg}], `
            // }
            else
            {
                argStr += $"{args[i]}, ";
            }
        }

        return argStr.Length > 2
            ? argStr = argStr.Substring(0, argStr.Length - 2)
            : argStr
        ;
    }



    // const level = {
    // pass: 10,
    // honour: 13,
    // }

    // const feedback = {
    // 0: `Try to earn 1 point to get started... 😏`,
    // 1: `Nice work, lets go, do another! 😉`,
    // 5: `Great, you are on fire and progress is made. 😋`,
    // 9: `Just one more to PASS. Lets go. 😅`,
    // 10: `Excellent, you have PASSED. Lets try one more? 😁`,
    // 13: `That is the way, you PASSED WITH HONOUR! 😍`,
    // 16: `What can I say. You impress me. 🙌`,
    // }

    // const stats = {}

    // /**
    // * Execute the testcase and assert that it was corrent or not and return a
    // * status string the can be written out.
    // * @param {Function} func - The function to test.
    // * @param {Object} expected - The expected result from the function call.
    // * @param {Object} args - The arguments to send to the function.
    // * @param {number} point - The number of points that this functions is worth in total.
    // * @returns {string} A string representation of the test case success or failure.
    // */
    // export function assert (func, expected, args, point=1) {
    // const result = func(...args)
    // const success = expected === result ? '✅' : '❌'

    // // Calculate the stats
    // if (!(func.name in stats)) {
    //     stats[func.name] = {
    //     run: 0,
    //     passed: 0,
    //     failed: 0,
    //     point,
    //     }
    // }
    // stats[func.name].run++
    // if (expected === result) {
    //     stats[func.name].passed++
    // } else {
    //     stats[func.name].failed++
    // }

    // // Prepare the argument string
    // let argStr = argsAsString(args)
    // let expectedStr = argsAsString([expected])
    // let resultStr = argsAsString([result])

    // return `${success} ${stats[func.name].point}p ${func.name}(${argStr}), expected: ${expectedStr} (${typeof expected}), actual: ${resultStr} (${typeof result})`
    // }

    // /**
    // * Print out the results when the execution is done.
    // * @returns {string} A string representation of the results.
    // */
    // export function done (func, expected, args) {
    // let points = 0
    // let passed = 0
    // let failed = 0
    // let total = 0
    // let result = 0
    // let str = ''
    // for (const func of Object.keys(stats)) {
    //     points += stats[func].point
    //     passed += stats[func].passed
    //     failed += stats[func].failed
    //     total += stats[func].passed + stats[func].failed
    //     result += stats[func].point * Math.floor(stats[func].passed / (stats[func].passed + stats[func].failed), 1)
    // }

    // let msg = feedback[0]
    // for (const key of Object.keys(feedback)) {
    //     if (result < key) {
    //     break
    //     }
    //     msg = feedback[key]
    // }

    // str += `Passed: ${passed}, Failed: ${failed}, Total: ${total}`
    // str += `\nYou have ${result} points out of total ${points} points.`
    // str += `\nYou need ${level.pass} points to PASS and ${level.honour} points to PASS WITH HONOUR.`
    // str += `\n${msg}`
    // return str
    // }
}

File: src/content/docs/laromaterial/lab_/orig/lab_java/lab_01/Lab.java
/**
 * This module contains lab specific code to assert the functions of the lab 
 * and to calculate the result.
 */
import java.lang.reflect.Method;
import java.util.LinkedHashMap;
import java.util.Map;

public class Lab {
    /**
     * Define the levels to PASS, PASS W HONOUR, PASS TOTAL.
     */
    static int PASS = 15;
    static int PASS_W_HONOUR = 19;
    static int PASS_TOTAL = 21;

    /**
     * Return a string with motivational feedback.
     * @returns {string} A string with a message.
     */
    static public String feedback (int points)
    {
        String msg = "Great, you are on fire and progress is made. 😋";
        if (points == 0) 
        {
            msg = "Try to earn 1 point to get started... 😏";
        }
        else if (points == 1)
        {
            msg = "Nice work, lets go, do another! 😉";
        }
        else if (points == PASS - 3)
        {
            msg = "Just three more to PASS. Lets go. 😅";
        }
        else if (points == PASS - 2)
        {
            msg = "Just two more to PASS. Lets go. 😅";
        }
        else if (points == PASS - 1)
        {
            msg = "Just one more to PASS. Lets go. 😅";
        }
        else if (points == PASS)
        {
            msg = "Excellent, you have PASSED. One more? 😁";
        }
        else if (points == PASS_W_HONOUR - 2)
        {
            msg = "Two more to PASS WITH HONOUR! Lets go. 😅";
        }
        else if (points == PASS_W_HONOUR - 1)
        {
            msg = "One more to PASS WITH HONOUR! Lets go. 😅";
        }
        else if (points == PASS_W_HONOUR)
        {
            msg = "That is the way, you PASSED WITH HONOUR! 😍";
        }
        else if (points == PASS_TOTAL)
        {
            msg = "What can I say. You impress me. 🙌";
        }

        return msg;
    }



    /**
     * Dictionary to hold statistics of each method call.
     */
    static Map<String, Statistic> stats = new LinkedHashMap<>();

    /**
    * Execute the testcase and assert that it was corrent or not and return a
    * status string the can be written out.
    * @param {Function} func - The function to test.
    * @param {Object} expected - The expected result from the function call.
    * @param {Object} args - The arguments to send to the function.
    * @param {number} point - The number of points that this functions is worth in total.
    * @returns {string} A string representation of the test case success or failure.
    */
    static public String assertIt(Method method, Object[] args, Object expected) {
        return assertIt(method, args, expected, 1);
    }

    static public String assertIt(
        Method method,
        Object[] args,
        Object expected,
        int point
    ) {
        // Call the method with the arguments
        Object result = null;
        try {
            result = method.invoke(null, args);
        } catch (Exception ex) {
            System.err.println(ex);
        }
        boolean success = false;

        // Verify that returned result matches the expected
        if (expected == null && result == null)
        {
            success = true;
        }
        else if (expected != null && result != null && expected.getClass() == result.getClass())
        {
            if (expected.getClass() == String.class && ((String)expected).equals((String)result))
            {
                success = true;
            }
            else if (expected.getClass() == Integer.class && (int)expected == (int)result)
            {
                success = true;
            }
            else if (expected.getClass() == Double.class && (double)expected == (double)result)
            {
                success = true;
            }
            else if (expected.getClass() == Boolean.class && (boolean)expected == (boolean)result)
            {
                success = true;
            }
        }

        // Calculate the stats
        String methodName = method.getName();
        if (stats.containsKey(methodName) == false)
        {
            stats.put(methodName, new Statistic(point));
        }
        stats.get(methodName).run(success);

        // Prepare the argument string
        String successStr = success ? "✅" : "❌";
        String points = String.valueOf(stats.get(methodName).points);
        String argStr = argsAsString(args);
        String expectedStr = expected != null ? argsAsString(new Object[]{expected}) : "";
        String resultStr = result != null ? argsAsString(new Object[]{result}) : "";
        String expectedType = expected != null ? expected.getClass().toString().substring(6) : "";
        String resultType = result != null ? result.getClass().toString().substring(6) : "";

        return String.format("%s %sp. %s(%s), expected: %s (%s), actual: %s (%s)", successStr, points, methodName, argStr, expectedStr, expectedType, resultStr, resultType);
    }



    /**
    * Format a value to its string representation.
    * @param {Object} args - A value to parse and format as a string.
    * @returns {string} A string representation of the value.
    */
    static String argsAsString(Object[] args)
    {
        String argStr = "";
        for (int i = 0; i < args.length; i++)
        {
            if (args[i].getClass() == String.class)
            {
                argStr += "\"" + args[i] + "\", ";
            }
            // else if (typeof arg === 'object' && Array.isArray(arg)) {
            //     argStr += `[${arg}], `
            // }
            else
            {
                argStr += args[i] + ", ";
            }
        }

        return argStr.length() > 2
            ? argStr = argStr.substring(0, argStr.length() - 2)
            : argStr
        ;
    }



    /**
    * Print out the results when the execution is done.
    * @returns {string} A string representation of the results.
    */
    static public String done ()
    {
        int passed = 0;
        int failed = 0;
        int total = 0;
        int result = 0;
        String str = "";
        boolean[] pointArray = new boolean[PASS_TOTAL];

        int step = 0;
        for (Statistic value : stats.values()) {
            passed += value.passed;
            failed += value.failed;
            total  += value.passed + value.failed;
            result += value.failed > 0 ? 0 : value.points;
            for (int i = 0; i < value.points; i++) {
                pointArray[step++] = value.failed == 0;
            }
        };

        String pointStr = "";
        for (int i = 1; i <= PASS_TOTAL; i++) {
            if (pointArray[i - 1]) {
                pointStr += " ⦿ ";
            } else if (i == PASS) {
                pointStr += " 😁 ";
            } else if (i == PASS_W_HONOUR) {
                pointStr += " 😍 ";
            } else if (i == PASS_TOTAL) {
                pointStr += " 🙌 ";
            } else {
                pointStr += " ⦾ ";
            }
        }

        str += " --------------------------------------------------------------------\n";
        str += "| Total: " + total + ", Passed ✅: " + passed + ", Failed ❌: " + failed;
        str += "\n|";
        str += "\n| Points needed to PASS=" + PASS + ", PASS WITH HONOUR=" + PASS_W_HONOUR + ", TOTAL=" + PASS_TOTAL;
        str += "\n| " + pointStr;
        str += "\n|\n| You have " + result + " points. " + feedback(result);
        str += "\n --------------------------------------------------------------------";

        return str;
    }
}



/**
 * Class to hold statistics summary for each method.
 */
class Statistic {
    public int run = 0;
    public int passed = 0;
    public int failed = 0;
    public int points;

    public Statistic (int aPoint)
    {
        points = aPoint;
    }

    public void run (boolean success)
    {
        run++;
        if (success) 
        {
            passed++;
        } 
        else
        {
            failed++;
        }
    }
}

